reputation/reputation.go

```diff
@@ -19,39 +25,165 @@ type ReputationModule struct {
 	rm := &ReputationModule{
 		ds: ds,
+		mi: mi,
+		si: si,
+		ai: ai,
 
+		rebuild: make(chan struct{}, 1),
 		ctx:     ctx,
 		cancel:  cancel,
 		sources: source.NewStore(ds),
 	}
 
 	go rm.updateSources()
+	go rm.subscribeIndexes()
+	go rm.indexBuilder()
 
 	return rm
 }
 
+func (rm *ReputationModule) AddSource(id string, maddr ma.Multiaddr) error {
+	return rm.sources.Add(source.Source{Id: id, Maddr: maddr})
+}
+
+func (rm *ReputationModule) GetTopMiners(n int) ([]MinerScore, error) {
+	mr := make([]MinerScore, 0, n)
+	rm.lockScores.Lock()
+	for i := 0; i < n; i++ {
+		mr = append(mr, rm.scores[i])
+	}
+	rm.lockScores.Unlock()
+	return mr, nil
+}
+
 func (rm *ReputationModule) Close() error {
 	rm.cancel()
 	return nil
 }
 
-func (rm *ReputationModule) AddSource(id string, maddr ma.Multiaddr) error {
-	return rm.sources.Add(source.Source{Id: id, Maddr: maddr})
+func (rm *ReputationModule) subscribeIndexes() {
+	subMi := rm.mi.Listen()
+	subSi := rm.si.Listen()
+	subAi := rm.ai.Listen()
+
+	for {
+		rm.lockIndex.Lock()
+		select {
+		case <-rm.ctx.Done():
+			log.Info("terminating background index update")
+			rm.lockIndex.Unlock()
+			return
+		case <-subMi:
+			rm.mIndex = rm.mi.Get()
+		case <-subSi:
+			rm.sIndex = rm.si.Get()
+		case <-subAi:
+			rm.aIndex = rm.ai.Get()
+		}
+		rm.lockIndex.Unlock()
+		select {
+		case rm.rebuild <- struct{}{}:
+		default:
+		}
+	}
+}
+
+func (rm *ReputationModule) indexBuilder() {
+	for range rm.rebuild {
+		log.Debug("rebuilding index")
+		start := time.Now()
+
+		sources, err := rm.sources.GetAll()
+		if err != nil {
+			log.Errorf("error when getting sources: %s", err)
+			return
+		}
+		rm.lockIndex.Lock()
+		minerIndex := rm.mIndex
+		slashIndex := rm.sIndex
+		askIndex := rm.aIndex
+		rm.lockIndex.Unlock()
+
+		scores := make([]MinerScore, 0, len(minerIndex.Miners))
+		for addr := range minerIndex.Miners {
+			score := calculateScore(addr, minerIndex, slashIndex, askIndex, sources)
+			scores = append(scores, score)
+		}
+		sort.Slice(scores, func(i, j int) bool {
+			return scores[i].Score > scores[j].Score
+		})
+
+		rm.lockScores.Lock()
+		rm.scores = scores
+		rm.lockScores.Unlock()
+
+		log.Debug("index rebuilt int %dms", time.Since(start).Milliseconds())
+	}
+}
+
+func calculateScore(addr string, mi miner.Index, si slashing.Index, ai ask.Index, ss []source.Source) MinerScore {
+	minfo := mi.Miners[addr]
+	powerScore := minfo.Power.Relative
+
+	var slashScore float64
+	if si, ok := si.Miners[addr]; ok {
+		slashScore = 1 / math.Pow(2, float64(len(si.History)))
+	}
+
+	var externalScore float64
+	for _, s := range ss {
+		score, exist := s.Scores[addr]
+		if !exist {
+			continue
+		}
+		externalScore = s.Weight * float64(score)
+	}
+
+	var askScore float64
+	if a, ok := ai.Miners[addr]; ok && a.Price < ai.MedianPrice {
+		askScore = 1
+	}
+
+	score := 50*slashScore + 20*powerScore + 20*externalScore + 10*askScore
+	return MinerScore{
+		Addr:  addr,
+		Score: int(score),
+	}
```

Comment:
Ok, this is just an attempt to do _something_. It doesn't mean is a meaningful tuned thing.
It uses information from the different indexes to compute a weighted average to come up with a number:
`score := 50*slashScore + 20*powerScore + 20*externalScore + 10*askScore`
In that formula, slashes have more weight than a good ask score, etc.

Pretty hard to know now what would be a good formula. The main idea is showing that all this reputation logic can be abstracted from all the Lotus logic and just gather all information availabe in local indexes and external `Source`s, to come up with some sorted list.

Reply:
There're more information in indexes, and it will have more the future. I just used something here in the formulas.

Reply:
Sounds like a great starting place... maybe PL will have some ideas about what makes for a good score algo.